### ISSUE TITLE
Issues running Godot demos

### ISSUE THREAD

**Issue by Maxim-Lanskoy** [migueldeicaza/SwiftGodotKit#20]
Hi @migueldeicaza! Thanks for you work and effort in providing new possibilities for Swift devs. I'm regular iOS dev and I've accidentally saw your live presentation at #GodotCon2023 and was really amazed with what you've developed here.

Everything seems clear for me during tutorials from documentation site and examples running, but I've struggled to go further. As a regular lazy dev I decided to start with some pre-made demo, and migrate it to Swift part by part to start something with. I've tried next demos, which are perfectly running through regular Godot Editor:

- ([Godot](https://github.com/godotengine)) [Third Person Shooter Demo](https://github.com/godotengine/tps-demo);
- ([GDQuest](https://github.com/gdquest-demos)) [RoboBlast: Third-Person Shooter](https://github.com/gdquest-demos/godot-4-3d-third-person-controller);
- ([Godot](https://github.com/godotengine)) [Platformer 3D](https://github.com/godotengine/godot-demo-projects/tree/master/3d/platformer);
- ([ywmaa](https://godotengine.org/asset-library/asset?user=ywmaa)) [Advanced Movement System Godot (AMSG) Template](https://godotengine.org/asset-library/asset/1107).

For all of them I've used the same setup. First, I've created four Swift executable packages, named _SwiftGodotTest_, in which I've added SwiftGodot and SwiftGodotKit as dependencies. Then, I've placed four mentioned Godot demo projects (contents of the root project directory, including _project.godot_ files) into _Sources/SwiftGodotTest_ folder of my Swift packages. Resulting packages are looking similar to this:
<img width="1080" alt="Screenshot 2024-01-04 at 04 02 36" src="https://github.com/migueldeicaza/SwiftGodotKit/assets/16775231/bf6a5a54-b5fd-46fc-9635-754a896c7bdd">
Here you can see package structure and that it consists only from single main Swift file and regular Godot project files.

For all demos scenes are loaded without errors and game window appears, but all of them were having issues at runtime:
- Unresponsive mouse - mouse was not tracked when game window appeared, until command-tab out-into back to the app + errors about CursorUIView Service appearing in console (`FAULT: <NSRemoteView: 0x11f4192b0 com.apple.TextInputUI.xpc.CursorUIViewService TUICursorUIViewService> determined it was necessary to configure <TUINSWindow: 0x11f4164f0> to support remote view vibrancy`);
- Irregular crash on camera movements (`[MTLDebugRenderCommandEncoder setRenderPipelineState:]:1615: failed assertion Set Render Pipeline State Validation
The color sample count (1) does not match the renderPipelineState's color sample count (4)
The raster sample count (1) does not match the renderPipelineState's raster sample count (4)`);
- Surface index out of bounds (`ERROR: Index (uint32_t)p_surface = 0 is out of bounds (mesh->surface_count = 0).
at: mesh_surface_set_material (servers/rendering/renderer_rd/storage_rd/mesh_storage.cpp:522)`);
- Vertex data mismatch (`Size of vertex data provided (27816) does not match expected (41724) at: mesh_add_surface`);
- Same visual issue for all demos with rendering ([video](https://github.com/migueldeicaza/SwiftGodotKit/assets/16775231/a4fb8efb-a627-4042-a834-7d434d1d1b3d)):
<img width="1080" alt="Screenshot 2024-01-04 at 04 39 26" src="https://github.com/migueldeicaza/SwiftGodotKit/assets/16775231/cc21b098-6b4b-4772-9f6c-c368f1d4b9d8">

Sorry if my question is not directly related to your Kit development, but is it supported to do such things at all? I mean running Godot projects with GDScripts via SwiftGodotKit package? Or there should be only Swift written files? Shouldn't running from Godot Project file and SwiftGodotKit be equal by result? I will be very thankful if you are able to provide me some ideas or directions in which I can investigate this further.

Firstly I thought, maybe some internal Godot settings are missing when running from Swift package, but then I've dealt with [settings callback issue](https://github.com/migueldeicaza/SwiftGodotKit/issues/10) and found out that if I placed _project.godot_ file at package root, settings already should be there, which is true, since I can see difference if I change something through Godot Editor (app's icon for example).

Thanks for your time and patience in advance.

**Comment by migueldeicaza** [comment]
The libgodot that I had published was an old version, let me update to the 4.2 version.

**Comment by Maxim-Lanskoy** [comment]
Thanks for your reply and provided update! Also I've found your Dodge project port in progress and take some helpful ideas from there.

I've rebuild all examples with new package versions and now they are running with much less issues.
The only blocking issue I'm facing right now, that if running a scene from Xcode, where any Spotlight source is present, process will crash because of error:
```
-[MTLDebugRenderCommandEncoder setRenderPipelineState:]:1615: failed assertion `Set Render Pipeline State Validation
The color sample count (1) does not match the renderPipelineState's color sample count (4)
The raster sample count (1) does not match the renderPipelineState's raster sample count (4)
```
If I remove all such light sources from any Godot demo or run it from Godot Editor itself, it will run without any blocking issues. Can we find out what Editor is doing to prevent such issue in runtime?

**Comment by migueldeicaza** [comment]
Lovely news!

The Dodge sample still has a few issues, as I changed a lot of variable names and they still don’t all match.

Godot has three different rendering engines/modes, it’s possible they are using different ones (it’s the thing you set when you create a new project, and the top text on the editor).

could you check that?

**Comment by Maxim-Lanskoy** [comment]
Yes, seems like we are looking into the right direction. I've changed rendering mode to "Compatibility" value from Godot Editor, saved project and then launched it from my Swift executable. Now I can use SpotLight3D sources without crash.
Quality of graphics is much worse though. It is possible to achieve "Forward" rendering mode working too? 

Anyway, the issue is not blocking anymore and I can continue development. Thanks for your help! Should I close this issue now or you want to run some more tests to support Forward rendering mode in nearest time?

**Comment by migueldeicaza** [comment]
These are issues generally with godot, and I believe they are working on those.

Additionally, there is a native Metal backend (instead of the emulated Vulkan-on-metal) coming, which might help.

Happy things worked out!
